File: fourthApp/fourthApp/game/common/src/main/scala/com/glyph/libgdx/util/LinkedList.java
package com.glyph.libgdx.util;

import java.util.Iterator;

/**
 * @author glyph
 */
public class LinkedList<E> implements Iterable<E>{
    Element head = new Element();
    Element tail = head;

    public class Iterator_ implements Iterator<E>{
        Element prev = head();
        Element current = prev;
        @Override
        public boolean hasNext() {
            boolean result = current.next != null;
            return result;
        }

        @Override
        public E next(){
            prev = current;
            current = current.next;
            return prev.element;
        }

        @Override
        public void remove() {
            prev.next = current.next;
        }
    };
    public void push(E e){
        Element next = new Element();
        next.element = e;
        tail.next = next;
    }

    public Element head(){
        return head;
    }

    public class Element{
        public E element;
        public Element next;
    }

    @Override
    public Iterator<E> iterator() {
        return new Iterator_();
    }
}

File: fourthApp/fourthApp/game/common/src/main/scala/com/glyph/scala/lib/util/collection/LinkedList.scala
package com.glyph.scala.lib.util.collection

import collection.mutable

/**
 * @author glyph
 */
class LinkedList[T] extends mutable.Traversable[T] {
  val mHead = new Element

  def push(e: T) {
    val next = new Element
    next.data = e
    next.next = mHead.next
    mHead.next = next
  }

  def pop(): T = {
    val result = mHead.next
    mHead.next = result.next
    val data = result.data
    data
  }


  def remove(t: T) {
    var current = mHead.next
    var prev = mHead
    var continue = current.next != null
    while (continue) {
      if (current.data == t) {
        prev.next = current.next
        continue = false
      } else {
        prev = current
        current = current.next
      }
    }
  }

  override def foreach[U](f: (T) => U) {
    var current = mHead
    while (current != null && current.next != null) {
      current = current.next
      f(current.data)
    }
  }

  def foreachPop[U](f: (T) => U) {
    while (!isEmpty) {
      f(pop())
    }
  }

  def clear() {
    while (!isEmpty) {
      pop()
    }
  }

  class Element {
    var data: T = null.asInstanceOf[T]
    var next: Element = null
  }

  override def isEmpty: Boolean = {
    mHead.next == null
  }
}
